subsys/net/lib/sockets/socketpair.c

```diff
@@ -401,17 +402,15 @@ static ssize_t spair_write(void *obj, const void *buffer, size_t count)
 			res = -1;
 			goto out;
 		}
+		is_nonblock = sock_is_nonblock(spair);
```

Comment:
Others know this mechanism better, but what's stopping sock_is_nonblock() from changing below this line? I'm not sure I understand how this closes the race. You have no synchronization object to prevent the nonblocking state from changing by some other thread (either preemption or running on another CPU) or ISR (either preempting this thread or firing on another CPU, if this state can be changed from ISRs).

Now in this test case it might not matter, this is more of a question as to what is the intended use of this API to begin with, since it never gives you *any* guarantees that the answer it gives you won't change at any time?

In other words, I don't think your change here fixes anything and it would probably be best to just drop this patch and leave this alone, since at least for this test case there will not be other threads or ISRs running that change the nonblocking state.

Reply:
I can understand why it is a bit confusing.

To answer your first question though, taking the semaphore on 399 prevents any other thread from changing the nonblocking state below 399 (and subsequently 405).

`spair->sem` is the synchronization object.

The confusing part is that you can't block on taking it the first time in case the socket is non-blocking. So if the socket is non-blocking, *and* you fail to take the semaphore, you know _with certainty_ that you must error and set EAGAIN.

Does that clarify the rest?

Reply:
I  think I see it now. Thanks for the extra detail!

Reply:
> Others know this mechanism better, but what's stopping sock_is_nonblock() from changing below this line?

> In other words, I don't think your change here fixes anything and it would probably be best to just drop this patch and leave this alone

A few words from me. I heard there's somethings scribbled on the stone "Thou shalt not share I/O resource among threads", so whoever does,  is looking for adventure. Of course, we should not open up security holes accessible with this, or crash badly (which is considered a security issue itself nowadays). I would conjecture that standard TOCTOU mitigation measure - make an important state copy at the beginning, consistently use the copy after wards - would help here. Oh, if we really talk of "mitigation", that copy apparently should be marked as volatile, given modern compilers' desire to eat more and more into the C semantics.

Beyond that, I neither tried to understand the changes in detail, I base my approval on the fact that we had pretty detailed discussion of various issues with @cfriedt, and there're pretty extensive tests, so it should be ok (and we'll anyway need a few iterations to get it right). So, I'm just trying to make sure that organizationally, changes are properly structured, e.g. code change can be reverted without affecting tests (the good criteria to consider what can (==should) go into separate commits).

> I neither tried to understand the changes in detail

(Oh, I don't do that because how would it look if I deeply nitpicked into somebody's code, given that this work brought up some known issues and technical debt in "my" code, and that's what I should be looking into instead? ;-) ).

Reply:
> Oh, if we really talk of "mitigation", that copy apparently should be marked as volatile, given modern compilers' desire to eat more and more into the C semantics.

Btw @andrewboie, wonder if syscall vrfy handlers are safe in that regard ;-).

Reply:
> Btw @andrewboie, wonder if syscall vrfy handlers are safe in that regard ;-).

I'm not sure I understand what you mean here.

Reply:
> I'm not sure I understand what you mean here.

Well, roughly speaking, a vrfy handler with a structure like https://github.com/zephyrproject-rtos/zephyr/blob/master/kernel/msg_q.c#L92 . Suppose we'd want to validate something *inside* struct k_msgq *q. Per TOCTOU rules, reliable way to do that would be to make a copy first:

~~~
struct k_msgq s = *q;

validate(s.some_field);
impl(&s);
~~~

My concern is that oversmart modern compilers could optimize that to:

~~~
validate(q->some_field);
impl(q);
~~~

To guaranteedly preclude that, we'd need to do `volatile struct k_msgq s = *q;`. And I don't believe I ever saw `volatile` keyword in any vrfy function, hence my question.

Reply:
In the provided code snippet you have no idea whether q is even valid memory or not. For the situation you describe the best practice is to use the `z_user_from_copy()` API.

I'm not sure why you're linking `z_vrfy_k_msgq_alloc_init` that's not an example of this situation.

I don't know if compilers perform the optimization you cite, but some evidence of your claims would be helpful. Or are you just speculating?

In any event `z_user_to_copy()` and `z_user_from_copy()` are what we want people using, there is no chance of that happening if they are used and they further will later enable us to implement security features like SMEP and SMAP.

There is discussion of these considerations here: https://docs.zephyrproject.org/latest/reference/usermode/syscalls.html#verification-memory-access-policies

Reply:
> I'm not sure why you're linking z_vrfy_k_msgq_alloc_init that's not an example of this situation.

Just a result of a quick grep of a func which accept a struct in memory, I don't imply there's something wrong with that func.

> Or are you just speculating?

Yep, just speculating, given that we discussed a case of how we could protect from races of external state change behind our back.

> There is discussion of these considerations here: https://docs.zephyrproject.org/latest/reference/usermode/syscalls.html#verification-memory-access-policies

Thanks for responses, I don't think I did all the homework I wanted on userspace and syscalls (another reason why converting posix ops to syscall was still in backlog).